File: Easy/validate-subsequence/solution_1/validate-subsequence.py
# SOLUTION 1
# Validate Subsequence

# Complexity
# Average:  Time: O(n)  | Space: O(1)
# Worst:    Time:       | Space:
# where n is the length of the array

def isValidSubsequence(array, sequence):
  arrIdx = 0
  seqIdx = 0

  while arrIdx < len(array) and seqIdx < len(sequence):
    if array[arrIdx] == sequence[seqIdx]:
      seqIdx += 1
    arrIdx += 1
  return seqIdx == len(sequence)

File: Easy/validate-subsequence/solution_1/validate-subsequence.ts
// resolve TypeScript error - Cannot redeclare block-scoped variable.
export {};

// SOLUTION 1
// Validate Subsequence

// Complexity
// Average:  Time: | Space:
// Worst:    Time: | Space:

// O(n) time | O(1) space

export function isValidSubsequence(array: number[], sequence: number[]) {
  let arrIdx = 0;
  let seqIdx = 0;

  while (arrIdx < array.length && seqIdx < sequence.length) {
    if (array[arrIdx] === sequence[seqIdx]) seqIdx++;
    arrIdx++;
  }
  return seqIdx === sequence.length;
}
